### ISSUE TITLE
JNP fft has different result than a forward IRFFTN

### ISSUE THREAD

**Issue by sarahbton** [google/jax#8009]
```python
import jax.numpy as jnp
import numpy as np
a = jnp.array([2, 174.81344591, 0, 0])
a = jnp.fft.fftn(a)
B = jnp.square(a)
C = jnp.fft.fftn(B, B.shape)
print(C) #outputs: [1.6000000e+01+0.j 2.4414062e-04+0.j 1.2223896e+05+0.j 2.7970154e+03+0.j]
D = jnp.fft.irfftn(B, B.shape)
DD = np.array([D[zc] * (len(D)/1) for zc in range(0, len(D))])
print(DD) #outputs: [1.6000000e+01 2.7970154e+03 1.2223896e+05 2.4414062e-04]


```

Those two outputs should theoretically match, because I am reversing the normalization that is caused by the default backward normalization for the irfftn. I believe this has something to do with the nyquist frequency portion of the irfftn but I couldn't figure it out. Any ideas?

In addition, are there any plans to support the "forward" and "backward" np fft arguments? They were added in 1.2.

**Comment by jakevdp** [comment]
Hi - thanks for the report. To be honest, I don't follow the logic of why the two outputs should be the same. I double-checked that doing the same computation with `numpy` shows a similar mismatch – the details differ, but if you set JAX to X64 mode, it matches numpy almost exactly.

Perhaps the expectation that these should match is incorrect?

**Comment by jakevdp** [comment]
Thanks - I still don't entirely understand the issue. You say that the 64-bit computations work as expected, but I see similar qualitative behavior to the 32-bit version, and this matches the numpy 64-bit version. To be clear, is the following the behavior you expect to see?
```python
from jax import config; config.update('jax_enable_x64', True)

import jax.numpy as jnp
import numpy as np

# JAX 64-bit
a = jnp.array([2, 174.81344591, 0, 0])
a = jnp.fft.fftn(a)
B = jnp.square(a)
C = jnp.fft.fftn(B, B.shape)
print(C) #outputs: [1.60000000e+01+0.j 1.13686838e-12+0.j 1.22238963e+05+0.j 2.79701513e+03+0.j]
D = jnp.fft.irfftn(B, B.shape)
DD = jnp.array([D[zc] * (len(D)/1) for zc in range(0, len(D))])
print(DD) #outputs: [1.60000000e+01 2.79701513e+03 1.22238963e+05 1.13686838e-12]

# numpy 64-bit
a = np.array([2, 174.81344591, 0, 0])
a = np.fft.fftn(a)
B = np.square(a)
C = np.fft.fftn(B, B.shape)
print(C) #outputs: [1.60000000e+01+0.j 1.13686838e-12+0.j 1.22238963e+05+0.j 2.79701513e+03+0.j]
D = np.fft.irfftn(B, B.shape)
DD = np.array([D[zc] * (len(D)/1) for zc in range(0, len(D))])
print(DD) #outputs: [1.60000000e+01 2.79701513e+03 1.22238963e+05 1.13686838e-12]
```
And if so, what is is about the 32-bit results that is unexpected?

**Comment by sarahbton** [comment]
> Thanks - I still don't entirely understand the issue. You say that the 64-bit computations work as expected, but I see similar qualitative behavior to the 32-bit version, and this matches the numpy 64-bit version. To be clear, is the following the behavior you expect to see?
> 
> ```python
> from jax import config; config.update('jax_enable_x64', True)
> 
> import jax.numpy as jnp
> import numpy as np
> 
> # JAX 64-bit
> a = jnp.array([2, 174.81344591, 0, 0])
> a = jnp.fft.fftn(a)
> B = jnp.square(a)
> C = jnp.fft.fftn(B, B.shape)
> print(C) #outputs: [1.60000000e+01+0.j 1.13686838e-12+0.j 1.22238963e+05+0.j 2.79701513e+03+0.j]
> D = jnp.fft.irfftn(B, B.shape)
> DD = jnp.array([D[zc] * (len(D)/1) for zc in range(0, len(D))])
> print(DD) #outputs: [1.60000000e+01 2.79701513e+03 1.22238963e+05 1.13686838e-12]
> 
> # numpy 64-bit
> a = np.array([2, 174.81344591, 0, 0])
> a = np.fft.fftn(a)
> B = np.square(a)
> C = np.fft.fftn(B, B.shape)
> print(C) #outputs: [1.60000000e+01+0.j 1.13686838e-12+0.j 1.22238963e+05+0.j 2.79701513e+03+0.j]
> D = np.fft.irfftn(B, B.shape)
> DD = np.array([D[zc] * (len(D)/1) for zc in range(0, len(D))])
> print(DD) #outputs: [1.60000000e+01 2.79701513e+03 1.22238963e+05 1.13686838e-12]
> ```
> 
> And if so, what is is about the 32-bit results that is unexpected?

Sorry for the delay, I did some digging and think I found the root cause. 

IFFT and FFT are not the same equation. The IFFT has a 2 rather than -2. See https://numpy.org/doc/stable/reference/routines.fft.html for more info.

Because currently JAX only allows None for the normalization option this default sets the normalization to backward. In a sense the IRFFTN component of the example code (where the output of it is multiplied by len(d)/1) just gets rid of the normalization that the backward defaults too. The order that IRFFTN gives is the one that I need for my algorithm. In the example code the 2nd and final values are in the wrong order (they should be switched) so that the numpy forward IRFFTN has the same sample output value of the JAX "forward" irfftn.

I did some digging through the code and have made some progress on changing the default for IRFFTN to a forward version but I don't have a ton of experience with jaxlib, bazel, xla, etc. so I figured you might be able to help out a bit. You effectively just have to set the normalization to be 1 by default for the IRFFTN. Do you know what line I would change?

In the long run JAX should ship with support for the numpy normalization options that were added in 1.2.

**Comment by sarahbton** [comment]
Hi Jake,

I got it working on my end. Had to edit some of the JAX source code to have it be forward be default. I looks like this isn't a bug but rather a pretty key feature that is missing.

Thanks for the help.

**Comment by jakevdp** [comment]
Thanks - to be clear, could you describe the key feature that is missing? And feel free to re-open this to track development of that feature. Thanks!

**Comment by sarahbton** [comment]
The ability to specify the norm option (forward, backward, ortho, default is backward) in the FFT and IFFT functions.

**Comment by jakevdp** [comment]
Thank you, I've opened https://github.com/google/jax/issues/8100 to track this request.
